Feature/osgar xmlrpc

m3d opened the pull request:
This PR is for discussion about logging/patching XML RPC server necessary for `ROSProxy`. Patch of `xmlrpc.client.SafeTransport` was failing due to SSL error:

```
ERROR: test_rpc_client (osgar.drivers.test_logrpc.LogRPCTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "m:\git\osgar\osgar\drivers\test_logrpc.py", line 67, in test_rpc_client
    self.assertEqual(s.add(2, 3), 5)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1112, in __call__
    return self.__send(self.__name, args)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1452, in __request
    verbose=self.__verbose
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1154, in request
    return self.single_request(host, handler, request_body, verbose)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1166, in single_request
    http_conn = self.send_request(host, handler, request_body, verbose)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1279, in send_request
    self.send_content(connection, request_body)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\xmlrpc\client.py",
line 1309, in send_content
    connection.endheaders(request_body)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\http\client.py", li
ne 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\http\client.py", li
ne 1026, in _send_output
    self.send(msg)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\http\client.py", li
ne 964, in send
    self.connect()
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\http\client.py", li
ne 1400, in connect
    server_hostname=server_hostname)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\ssl.py", line 401,
in wrap_socket
    _context=self, _session=session)
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\ssl.py", line 808,
in __init__
    self.do_handshake()
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\ssl.py", line 1061,
 in do_handshake
    self._sslobj.do_handshake()
  File "D:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\lib\ssl.py", line 683,
in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: UNKNOWN_PROTOCOL] unknown protocol (_ssl.c:748)
```
which is maybe related to old Python only??

At the moment only reasonable/working experiment is with `patch` and `wraps`.

zwn commented:
Maybe we need to start with something like this?

```python
import types, enum

def wrap_namespace(old, new):
    for name, obj in old.items():
        t = type(obj)
        if t in (types.BuiltinFunctionType, str, int, bool, dict, list, set) or isinstance(obj, (enum.Enum, enum.EnumMeta)):
            continue
        if t is types.FunctionType:
            print("func: ", name)
        elif t is type:
            print("class: ", name)
        elif t is types.ModuleType:
            print("module: ", name)
        else:
            print("unknown:", name, t)
        new[name] = obj

import socket
wrap_namespace(socket.__dict__, globals())
````

zwn commented:
For me it returns:
```
unknown: __loader__ <class '_frozen_importlib_external.SourceFileLoader'>
unknown: __spec__ <class '_frozen_importlib.ModuleSpec'>
module:  _socket
func:  socketpair
func:  getaddrinfo
class:  error
class:  herror
class:  gaierror
class:  timeout
class:  SocketType
class:  socket
unknown: CAPI <class 'PyCapsule'>
module:  os
module:  sys
module:  io
module:  selectors
module:  errno
func:  _intenum_converter
class:  _realsocket
class:  _GiveupOnSendfile
func:  fromfd
unknown: SocketIO <class 'abc.ABCMeta'>
func:  getfqdn
unknown: _GLOBAL_DEFAULT_TIMEOUT <class 'object'>
func:  create_connection
```
If we can create a function wrapper and class wrapper we might be good to go (?).

zwn commented:
On idiomatic function wrappers: https://stackoverflow.com/a/45027988/3185929

zwn commented:
[Capture-Replay Mocking with CaptureMock](http://texttest.sourceforge.net/index.php?page=capturemock&n=python_basic)

zwn commented:
In case someone would want to continue,  here is some WIP:
```python
def wrap_func(func):
    @wraps(func)
    def wrapped(*args, **kwargs):
        print("pre")
        ret = func(*args, **kwargs)
        print("post")
        return ret
    return wrapped

def wrap_class(klass):
    class Wrapped:
        def __init__(self, *args, **kwargs):
            print("initing")
            self._orig = klass(*args, **kwargs)
        def __getattr__(self, attr):
            orig_attr = getattr(self._orig, attr)
            if (isinstance(orig_attr, types.MethodType)):
                print("wrapping")
                wrapped = wrap_func(orig_attr)
                setattr(self, attr, wrapped)
                return wrapped
            return orig_attr
    return Wrapped
```
And the tests:
```python
def test_func(name):
    return "Hello " + name + "!"

class test_class:
    def __init__(self, test_arg):
        self.test_arg = test_arg
    def method(self):
        return self.test_arg
        
print("---- test function wrapping ----")
a = test_func("robot")
test_func_wrapped = wrap_func(test_func)
b = test_func_wrapped("robot")
print("result: ", b)
assert a == b

print("---- test class wrapping ----")
A = test_class("robot")
test_class_wrapped = wrap_class(test_class)
B = test_class_wrapped("robot")
print("result: ", B.method())
assert A.method() == B.method()

print("---- test accessing values ---")
print("result: ", A.test_arg)
assert A.test_arg == B.test_arg
```
And the result of the tests:
```
---- test function wrapping ----
pre
post
result:  Hello robot!
---- test class wrapping ----
initing
wrapping
pre
post
result:  robot
pre
post
---- test accessing values ---
result:  robot
```

m3d commented:
thanks for the code - I added it to experimental tests, fixed imports and it worked fine, but ...
when I use it for `xmlrpc.client.ServerProxy` it fails. Is it expected? I suppose that the proxy creates functions "on demand" (???). If you use
```
s = wrap_class(xmlrpc.client.ServerProxy('http://localhost:8000'))
```
then it fails
```
M:\git\osgar>python -m unittest osgar\drivers\test_logrpc.py
E
======================================================================
ERROR: test_rpc_client_wraps (osgar.drivers.test_logrpc.LogRPCTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "M:\git\osgar\osgar\drivers\test_logrpc.py", line 149, in test_rpc_client
_wraps
    self.assertEqual(s.add(2, 3), 5)
AttributeError: type object 'Wrapped' has no attribute 'add'

----------------------------------------------------------------------
Ran 1 test in 0.006s

FAILED (errors=1)
```
???

zwn commented:
You need to wrap the class itself, not its instance. Please take a look at my test code and note the difference.

zwn commented:
Instead of 
```python
s = wrap_class(xmlrpc.client.ServerProxy('http://localhost:8000'))
```
you need to do
```python
wrappedServerProxyClass = wrap_class(xmlrpc.client.ServerProxy)
instance = wrappedServerProxyClass('http://localhost:8000')
```
But it may fail anyway if the class being wrapped does something out of ordinary. In any case the implementation can be adjusted to account for it once we need it to.

m3d commented:
Maybe the main blocker was non-functional `xmlrpc.client.SafeTransport` (probably solvable with some upgrades), and we could use workaround with `xmlrpc.client.Transport` ... and use your `wrap_class` for that? :)

m3d commented:
current output:
```
M:\git\osgar>python -m unittest osgar\drivers\test_logrpc.py
send_request localhost:8000 /RPC2 b"<?xml version='1.0'?>\n<methodCall>\n<method
Name>add</methodName>\n<params>\n<param>\n<value><int>2</int></value>\n</param>\
n<param>\n<value><int>3</int></value>\n</param>\n</params>\n</methodCall>\n" Fal
se
make connection localhost:8000
initing
127.0.0.1 - - [19/Feb/2019 07:16:22] "POST /RPC2 HTTP/1.1" 200 -
.
----------------------------------------------------------------------
Ran 1 test in 1.014s

OK
```

zwn commented:
To tell the truth, I don't have a working ROS environment yet so I cannot do any real development regarding ROS. I also know very little about the implementation of xmlrpc in python and all the classes involved so I hesitate to give any comments on that as well. I tried to help with generic wrapping functions and classes as that is something I have experience with.

So if there is something specific that should be achieved within this PR, could you please reformulate the task again?

m3d commented:
> 
> 
> To tell the truth, I don't have a working ROS environment yet so I cannot do any real development regarding ROS. I also know very little about the implementation of xmlrpc in python and all the classes involved so I hesitate to give any comments on that as well. I tried to help with generic wrapping functions and classes as that is something I have experience with.
> 
> So if there is something specific that should be achieved within this PR, could you please reformulate the task again?

OK, communication with ROS is based on XMLRPC and communication via TCP sockets. We already have logging of TCP sockets, so this PR was supposed to fill the gap of logging communication between OSGAR and ROS.

The status at the moment is that for `xmlrpc.client.ServerProxy` we need to log `xmlrpc.client.Transport`. It seems that override `send_request` and then parser used in `parse_response`.

The next step is to dynamically add TCP and RPC server/clients within `ROSProxy`.

p.s. you should get some ROS environment running anyway ... ;-)

m3d commented:
OK, I rebased it all to the last commit, which I would test within `ROSProxy`. At the moment it is logger only.

m3d commented:
p.s. I am looking at `d:\WinPython-64bit-3.6.2.0Qt5\python-3.6.2.amd64\Lib\xmlrpc\server.py` and the server is inherited from `socketserver.TCPServer`:
```
class SimpleXMLRPCServer(socketserver.TCPServer,
                         SimpleXMLRPCDispatcher):
```
so maybe all we need to do on server side is to use logged TCPServer??

Do you prefer to merge and then to try to integrate to `ROSProxy` or create another branch on top of this one? I would start with the client and then try server side later.
